sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala

```diff
@@ -0,0 +1,392 @@
+ * @param config  a collection of configuration options that will be added to the hive conf before
+ *                opening the hive client.
+ */
+class ClientWrapper(
+    version: HiveVersion,
+    config: Map[String, String])
+  extends ClientInterface
+  with Logging
+  with ReflectionMagic {
+
+  private val conf = new HiveConf(classOf[SessionState])
+  config.foreach { case (k, v) =>
+    logDebug(s"Hive Config: $k=$v")
+    conf.set(k, v)
+  }
+
+  private def properties = Seq(
+    "javax.jdo.option.ConnectionURL",
+    "javax.jdo.option.ConnectionDriverName",
+    "javax.jdo.option.ConnectionUserName")
+
+  properties.foreach(p => logInfo(s"Hive Configuration: $p = ${conf.get(p)}"))
+
+  // Circular buffer to hold what hive prints to STDOUT and ERR.  Only printed when failures occur.
+  private val outputBuffer = new java.io.OutputStream {
+    var pos: Int = 0
+    var buffer = new Array[Int](10240)
+    def write(i: Int): Unit = {
+      buffer(pos) = i
+      pos = (pos + 1) % buffer.size
+    }
+
+    override def toString: String = {
+      val (end, start) = buffer.splitAt(pos)
+      val input = new java.io.InputStream {
+        val iterator = (start ++ end).iterator
+
+        def read(): Int = if (iterator.hasNext) iterator.next() else -1
+      }
+      val reader = new BufferedReader(new InputStreamReader(input))
+      val stringBuilder = new StringBuilder
+      var line = reader.readLine()
+      while(line != null) {
+        stringBuilder.append(line)
+        stringBuilder.append("\n")
+        line = reader.readLine()
+      }
+      stringBuilder.toString()
+    }
+  }
+
+  val state = {
+    val original = Thread.currentThread().getContextClassLoader
+    Thread.currentThread().setContextClassLoader(getClass.getClassLoader)
+    val ret = try {
+      val newState = new SessionState(conf)
+      SessionState.start(newState)
+      newState.out = new PrintStream(outputBuffer, true, "UTF-8")
+      newState.err = new PrintStream(outputBuffer, true, "UTF-8")
+      newState
+    } finally {
+      Thread.currentThread().setContextClassLoader(original)
+    }
+    ret
+  }
+
+  private val client = Hive.get(conf)
+
+  private def withClassLoader[A](f: => A): A = synchronized {
+    val original = Thread.currentThread().getContextClassLoader
+    Thread.currentThread().setContextClassLoader(getClass.getClassLoader)
+    Hive.set(client)
+    version match {
+      case hive.v12 =>
+        classOf[SessionState]
+          .callStatic[SessionState, SessionState]("start", state)
+      case hive.v13 =>
+        classOf[SessionState]
+          .callStatic[SessionState, SessionState]("setCurrentSessionState", state)
+    }
+    val ret = try f finally {
+      Thread.currentThread().setContextClassLoader(original)
+    }
+    ret
+  }
+
+  def currentDatabase: String = withClassLoader {
+    state.getCurrentDatabase
+  }
+
+  def createDatabase(tableName: String): Unit = withClassLoader {
```

Comment:
something is wrong with this function. createDatabase takes a tableName, and has a local variable "table" that is not used.